🐛 Fix handling of logging with string and object arguments

pysysops opened the pull request:
closes #45
Fixes a couple of issues including: `logging.warn('A friendly message.', err);` throwing a TypeError and `logging.warn('A friendly message.', {err: err});` resulting in the error printed as: [object Object].

pysysops committed:
:bug: Fix handling of logging with string and object arguments

closes #45
Fixes a couple of issues including: `logging.warn('A friendly message.', err);` throwing a TypeError and `logging.warn('A friendly message.', {err: err});` resulting in the error printed as: [object Object].

pysysops commented:
This _should_ be good to go now. it rewrites the log function to handle objects and log messages separately then send them to bunyan as an exploded array ensuring that objects or errors are the first argument.

I've checked log output using this branch in Ghost and everything looks okay:
```
[2019-04-08 19:48:00] INFO "GET /ghost/api/v2/admin/slugs/post/a%20Post/" 200 18ms
[2019-04-08 19:48:00] INFO "POST /ghost/api/v2/admin/posts/" 201 97ms
[2019-04-08 19:48:07] INFO "PUT /ghost/api/v2/admin/posts/5cab97606123b41bc90135d5/" 200 89ms
[2019-04-08 19:48:07] INFO Trigger Webhook for  "site.changed" with url "https://<<REDACTED>>/".
[2019-04-08 19:48:31] INFO "GET /" 200 115ms
[2019-04-08 19:48:31] INFO "GET /assets/built/screen.css?v=6887c61d3c" 304 1ms
[2019-04-08 19:48:31] INFO "GET /assets/built/jquery.fitvids.js?v=6887c61d3c" 304 2ms
[2019-04-08 19:48:31] INFO "GET /assets/built/infinitescroll.js?v=6887c61d3c" 304 2ms
[2019-04-08 19:48:33] INFO "GET /a-post/" 200 83ms
[2019-04-08 19:48:33] INFO "GET /assets/built/screen.css?v=6887c61d3c" 304 2ms
[2019-04-08 19:48:33] INFO "GET /assets/built/jquery.fitvids.js?v=6887c61d3c" 304 3ms
[2019-04-08 19:48:37] INFO "GET /a-post/amp" 301 1ms
[2019-04-08 19:48:37] INFO amp.parse http://localhost:2368/a-post/ 4ms
[2019-04-08 19:48:37] INFO "GET /a-post/amp/" 200 161ms
```

naz commented on lib/logging/GhostLogger.js:
```diff
@@ -297,34 +297,30 @@ class GhostLogger {
+             if (value instanceof Error) {
+                 modifiedObject.err = value;
+             } else if (isObject(value)) {
+               var keys = Object.keys(value);
```
Something is off with formatting here and no need to declare additional variable I think, `each(Object.keys(value), function (key) {` is rather readable

naz commented on lib/logging/GhostLogger.js:
```diff
@@ -297,34 +297,30 @@ class GhostLogger {
+             modifiedMessages.push(modifiedObject.err.message);
+         }
+         
+         modifiedArguments.push(modifiedObject);
```
As Kate pointed out https://github.com/TryGhost/Ignition/pull/87#commitcomment-33085593, we should check if `_.isEmpty(modifiedObject)` here because we pollute the arguments with no need :wink:

pysysops committed:
Add missing gelf and name options

no issue

pysysops committed:
Improved log function

no issue
Cleaned up unnecessary var assignment, fixed indentation and added empty check for log objects

pysysops commented:
There's still something strange in the pretty streaming of errors with custom messages. The message doesn't get output as it should. I think it comes down to this section of code: https://github.com/TryGhost/Ignition/blob/master/lib/logging/PrettyStream.js#L132-L136

naz commented:
@pysysops if that section kicks in it means something is wrong with the data that is processed above - https://github.com/TryGhost/Ignition/blob/master/lib/logging/PrettyStream.js#L124-L131 (it's also a very strange way to control the flow through try/catch, would rather substitute that section with safe `_.get()` because try/catches are slow, which has a big effect in performance, especially when it comes to logging). Do you have some specific cases where the formatting is failing?

pysysops commented:
@gargol I'm looking at that whole section now. I have an initial rewrite of it. I think the try / catch can be replaced with an `if (data.req) ...` although I kinda feel there's too much control flow going on. But it can be improved iteratively. In reality the pretty stream / console logging should be used for nothing other than local development.

If we were to want to log to STDOUT to let something else handle it for example docker or systemd, we would probably want to send non-pretty logs as they would be parsed and presented by some other system.

pysysops commented:
As for an example, When we log something like: `logging.error('Something failed calling service x', err);` it hits this:
https://github.com/TryGhost/Ignition/blob/master/lib/logging/PrettyStream.js#L107

and eventually:
https://github.com/TryGhost/Ignition/blob/master/lib/logging/PrettyStream.js#L132-L137

and the message we have passed in is gone. I would expect the log to be:
```
[DATE HERE] ERROR Something failed calling service x

... ERROR DETAILS ... 
```
But currently you get:
```
[DATE HERE] ERROR

... ERROR DETAILS ... 
```

and potentially lose valuable context. Probably ending up with the error message / line numbers from a file in `node_modules`, if we output the msg at least it becomes grep'able.

pysysops committed:
Fixed PrettyStream output inconsistency

no issue

pysysops commented:
@gargol / @kirrg001 This is good for a final review. We should be able to merge it if everything looks okay?

naz commented on lib/logging/PrettyStream.js:
```diff
@@ -102,118 +102,108 @@ class PrettyStream extends Transform {
+                data.req.method.toUpperCase(),
+                data.req.originalUrl,
+                statusCode(data.res.statusCode),
+                data.res.responseTime
```
think it's worth putting before mentioned [safe getter](https://lodash.com/docs/4.17.11#get) around properties here, e.g:
```
_.get(data, 'req.originalUrl')
_.get(data, 'res.responseTime')
```
think that was the reason behind try/catch being here in the first place (specifically `data.res.responseTime` is very [Ghost specific field](https://github.com/TryGhost/Ghost/blob/fcd275f6c098684ad615b02d0714cf38e5cd8f44/core/server/web/shared/middlewares/log-request.js#L13))

naz commented:
@pysysops just one comment https://github.com/TryGhost/Ignition/pull/87/files#r273871797, the rest looks good :+1:

pysysops committed:
Add safe get() to data.res values

pysysops commented:
Resolved a merge conflict whilst rebasing hence force-push.
